Add wrapper code to wrap an ET kernel into an ATen kernel

larryliu0820 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #2186

Summary:

Add an API that is able to wrap ExecuTorch kernels into ATen kernels.

This can save kernel author time, if they start from ExecuTorch kernels.

In this PR we only support the following argument types to ExecuTorch
kernels:

* Tensor-like. `torch::executor::Tensor` and its reference/const
  reference.
* Primitive types. `int` `float` etc.

Doesn't support ArrayRef or Optional yet.

Test Plan:

Rely on unit tests

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D54394149](https://our.internmc.facebook.com/intern/diff/D54394149)

larryliu0820 committed:
Add wrapper code to wrap an ET kernel into an ATen kernel

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

larryliu0820 committed:
Update on "Add wrapper code to wrap an ET kernel into an ATen kernel"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

larryliu0820 committed:
Update on "Add wrapper code to wrap an ET kernel into an ATen kernel"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

larryliu0820 commented:
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D54394149).

larryliu0820 committed:
Update on "Add wrapper code to wrap an ET kernel into an ATen kernel"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D54394149](https://our.internmc.facebook.com/intern/diff/D54394149)

[ghstack-poisoned]

larryliu0820 committed:
Update on "Add wrapper code to wrap an ET kernel into an ATen kernel"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D54394149](https://our.internmc.facebook.com/intern/diff/D54394149)

[ghstack-poisoned]

larryliu0820 committed:
Update on "Add wrapper code to wrap an ET kernel into an ATen kernel"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D54394149](https://our.internmc.facebook.com/intern/diff/D54394149)

[ghstack-poisoned]

zou3519 commented on extension/aten_util/test/make_aten_functor_from_et_functor_test.cpp:
```diff
@@ -0,0 +1,106 @@
+}
+
+TORCH_LIBRARY(my_op, m) {
+  m.def("add_1.out", WRAP(add_1_out, 1));
```
Just to make sure I understand:
- we're saying that `add_1_out` is written in the executorch style? and WRAP makes it work with the torch style?

larryliu0820 replied:
Yes. `WRAP` allows `add_1_out` to take `at::Tensor` instead of `torch::executor::Tensor`

iseeyuan commented on extension/aten_util/make_aten_functor_from_et_functor.h:
```diff
@@ -0,0 +1,173 @@
+#define _WRAP_1(func) wrapper_impl<decltype(&func), func>::wrap
+
+#define GET_MACRO(_1, _2, NAME, ...) NAME
+#define WRAP(...) GET_MACRO(__VA_ARGS__, _WRAP_2, _WRAP_1)(__VA_ARGS__)
```
The name WRAP is too generic. Consider WRAP_TO_ATEN

iseeyuan reviewed (commented):
What types are mapped from ExecuTorch to ATen? Does it works for tensors, list<Tensors> and prim types? Please add what can be mapped and what cannot be mapped.

larryliu0820 committed:
Update on "Add wrapper code to wrap an ET kernel into an ATen kernel"


Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D54394149](https://our.internmc.facebook.com/intern/diff/D54394149)

[ghstack-poisoned]

larryliu0820 commented:
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D54394149).

larryliu0820 committed:
Update on "Add wrapper code to wrap an ET kernel into an ATen kernel"


Summary:

Add an API that is able to wrap ExecuTorch kernels into ATen kernels.

This can save kernel author time, if they start from ExecuTorch kernels.

Test Plan:

Rely on unit tests
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D54394149](https://our.internmc.facebook.com/intern/diff/D54394149)

[ghstack-poisoned]

larryliu0820 commented:
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D54394149).

larryliu0820 committed:
Update on "Add wrapper code to wrap an ET kernel into an ATen kernel"


Summary:

Add an API that is able to wrap ExecuTorch kernels into ATen kernels.

This can save kernel author time, if they start from ExecuTorch kernels.

In this PR we only support the following argument types to ExecuTorch
kernels:

* Tensor-like. `torch::executor::Tensor` and its reference/const
  reference.
* Primitive types. `int` `float` etc.

Doesn't support ArrayRef or Optional yet.

Test Plan:

Rely on unit tests

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D54394149](https://our.internmc.facebook.com/intern/diff/D54394149)

[ghstack-poisoned]

larryliu0820 commented:
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D54394149).

larryliu0820 committed:
Update on "Add wrapper code to wrap an ET kernel into an ATen kernel"


Summary:

Add an API that is able to wrap ExecuTorch kernels into ATen kernels.

This can save kernel author time, if they start from ExecuTorch kernels.

In this PR we only support the following argument types to ExecuTorch
kernels:

* Tensor-like. `torch::executor::Tensor` and its reference/const
  reference.
* Primitive types. `int` `float` etc.

Doesn't support ArrayRef or Optional yet.

Test Plan:

Rely on unit tests

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D54394149](https://our.internmc.facebook.com/intern/diff/D54394149)

[ghstack-poisoned]

larryliu0820 commented:
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D54394149).

larryliu0820 committed:
Update on "Add wrapper code to wrap an ET kernel into an ATen kernel"


Summary:

Add an API that is able to wrap ExecuTorch kernels into ATen kernels.

This can save kernel author time, if they start from ExecuTorch kernels.

In this PR we only support the following argument types to ExecuTorch
kernels:

* Tensor-like. `torch::executor::Tensor` and its reference/const
  reference.
* Primitive types. `int` `float` etc.

Doesn't support ArrayRef or Optional yet.

Test Plan:

Rely on unit tests

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D54394149](https://our.internmc.facebook.com/intern/diff/D54394149)

[ghstack-poisoned]

larryliu0820 commented:
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D54394149).

larryliu0820 committed:
Update on "Add wrapper code to wrap an ET kernel into an ATen kernel"


Summary:

Add an API that is able to wrap ExecuTorch kernels into ATen kernels.

This can save kernel author time, if they start from ExecuTorch kernels.

In this PR we only support the following argument types to ExecuTorch
kernels:

* Tensor-like. `torch::executor::Tensor` and its reference/const
  reference.
* Primitive types. `int` `float` etc.

Doesn't support ArrayRef or Optional yet.

Test Plan:

Rely on unit tests

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D54394149](https://our.internmc.facebook.com/intern/diff/D54394149)

[ghstack-poisoned]

larryliu0820 commented:
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D54394149).

larryliu0820 committed:
Update on "Add wrapper code to wrap an ET kernel into an ATen kernel"


Summary:

Add an API that is able to wrap ExecuTorch kernels into ATen kernels.

This can save kernel author time, if they start from ExecuTorch kernels.

In this PR we only support the following argument types to ExecuTorch
kernels:

* Tensor-like. `torch::executor::Tensor` and its reference/const
  reference.
* Primitive types. `int` `float` etc.

Doesn't support ArrayRef or Optional yet.

Test Plan:

Rely on unit tests

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D54394149](https://our.internmc.facebook.com/intern/diff/D54394149)

[ghstack-poisoned]

larryliu0820 commented:
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D54394149).

larryliu0820 commented:
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D54394149).

larryliu0820 committed:
Update on "Add wrapper code to wrap an ET kernel into an ATen kernel"


Summary:

Add an API that is able to wrap ExecuTorch kernels into ATen kernels.

This can save kernel author time, if they start from ExecuTorch kernels.

In this PR we only support the following argument types to ExecuTorch
kernels:

* Tensor-like. `torch::executor::Tensor` and its reference/const
  reference.
* Primitive types. `int` `float` etc.

Doesn't support ArrayRef or Optional yet.

Test Plan:

Rely on unit tests

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D54394149](https://our.internmc.facebook.com/intern/diff/D54394149)

[ghstack-poisoned]

larryliu0820 commented:
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D54394149).

larryliu0820 committed:
Update on "Add wrapper code to wrap an ET kernel into an ATen kernel"


Summary:

Add an API that is able to wrap ExecuTorch kernels into ATen kernels.

This can save kernel author time, if they start from ExecuTorch kernels.

In this PR we only support the following argument types to ExecuTorch
kernels:

* Tensor-like. `torch::executor::Tensor` and its reference/const
  reference.
* Primitive types. `int` `float` etc.

Doesn't support ArrayRef or Optional yet.

Test Plan:

Rely on unit tests

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D54394149](https://our.internmc.facebook.com/intern/diff/D54394149)

[ghstack-poisoned]

larryliu0820 commented:
@larryliu0820 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D54394149).

larryliu0820 committed:
Update on "Add wrapper code to wrap an ET kernel into an ATen kernel"


Summary:

Add an API that is able to wrap ExecuTorch kernels into ATen kernels.

This can save kernel author time, if they start from ExecuTorch kernels.

In this PR we only support the following argument types to ExecuTorch
kernels:

* Tensor-like. `torch::executor::Tensor` and its reference/const
  reference.
* Primitive types. `int` `float` etc.

Doesn't support ArrayRef or Optional yet.

Test Plan:

Rely on unit tests

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D54394149](https://our.internmc.facebook.com/intern/diff/D54394149)

[ghstack-poisoned]
